### DISCUSSION TITLE
System requirements documentation

### DISCUSSION THREAD

**Discussion by gritelli** [systemd/mkosi#2870] [General]
My mkosi builds are failing probably due to my Ubuntu and systemd versions. I have been looking for some documentation for system requirements to successfully run mkosi but I haven't found much. Does anybody know a good place to start for this?
How do I need to setup my Ubuntu to run mkosi?

**Comment by daandemeyer** [comment]
@gritelli `mkosi dependencies` will install all required packages to run mkosi. It just installs packages though, but they may still be too old depending on your distro and release. We need at least systemd v254. In other words, Jammy or earlier is too old, Noble will work.

Alternatively, you can use `--tools-tree` and mkosi will build a second image containing all the necessary tools and use that to build the final image.

**Reply by gritelli** [reply]
Got it. But I am running noble and systemd v255. See my apt list output below. Despite I'm still getting the following error:
`Could not find 'systemd-repart' which is required to build disk images.`

So do I run `mkosi --tools-tree` and it will generate an image that I can flash and then use to build new images?


_systemd-container/noble-updates,now 255.4-1ubuntu8.1 amd64 [installed,automatic]
systemd-dev/noble-updates,noble-updates,now 255.4-1ubuntu8.1 all [installed,automatic]
systemd-hwe-hwdb/noble,noble,now 255.1.3 all [installed,automatic]
systemd-oomd/noble-updates,now 255.4-1ubuntu8.1 amd64 [installed,automatic]
systemd-resolved/noble-updates,now 255.4-1ubuntu8.1 amd64 [installed]
systemd-sysv/noble-updates,now 255.4-1ubuntu8.1 amd64 [installed,automatic]
systemd-timesyncd/noble-updates,now 255.4-1ubuntu8.1 amd64 [installed,automatic]
systemd/noble-updates,now 255.4-1ubuntu8.1 amd64 [installed,automatic]_

**Reply by daandemeyer** [reply]
@gritelli You can use either `mkosi dependencies | xargs sudo apt-get install` or `mkosi --tools-tree` if you're on Noble.

`mkosi --tools-tree` will build an image and use it automatically to build the final image, no need to flash anything.

**Reply by gritelli** [reply]
Do you mind explaining these commands better?
I am running on Noble.

Say I have a `mkosi.conf` to build, do I do `mkosi build --tools-tree`?
I just made a fresh Ubuntu 24.04 installation, ran `sudo apt install mkosi`, and ran `sudo mkosi build --tools-tree` and this is what I get:
`‣ Building default image`  
`‣ Could not find 'systemd-repart' which is required to build disk images.`

Not sure how to move forward.
Thank you for your help

**Reply by daandemeyer** [reply]
@gritelli Options have to be passed before the verb, not afterwards. So you'll want to run `mkosi --tools-tree build`

**Reply by gritelli** [reply]
yes...but still no luck.

**Reply by daandemeyer** [reply]
@gritelli This does not help me, please paste full debug logs (`--debug`)

**Reply by gritelli** [reply]
actually if I run `mkosi --tools-tree build` I get this:
```
usage: 
  mkosi [options...] summary
  mkosi [options...] build       [command line...]
  mkosi [options...] shell       [command line...]
  mkosi [options...] boot        [nspawn settings...]
  mkosi [options...] qemu        [qemu parameters...]
  mkosi [options...] ssh         [command line...]
  mkosi [options...] journalctl  [command line...]
  mkosi [options...] coredumpctl [command line...]
  mkosi [options...] clean
  mkosi [options...] serve
  mkosi [options...] bump
  mkosi [options...] genkey
  mkosi [options...] documentation
  mkosi [options...] help
  mkosi -h | --help
  mkosi --version
mkosi: error: argument --tools-tree: expected one argument
```

**Reply by gritelli** [reply]
this is the output of `sudo mkosi --debug build`:

```
‣ Including configuration file /home/talos/Desktop/mkosi.conf
‣ Including configuration file /tmp/tmpwymtuct9/mkosi-tools/mkosi.conf
‣ Including configuration file /tmp/tmpwymtuct9/mkosi-tools/mkosi.conf.d/10-debian-ubuntu.conf
‣ Including configuration file /tmp/tmp9rdsz1ix/mkosi-tools/mkosi.conf
‣ Including configuration file /tmp/tmp9rdsz1ix/mkosi-tools/mkosi.conf.d/10-debian-ubuntu.conf
‣ + mount --make-rslave /
‣ + mount --rbind /usr /usr --options ro
‣ + mount --rbind /etc /etc --options ro
‣ + mount --rbind /opt /opt --options ro
‣ + mount --rbind /boot /boot --options ro
‣ + mount --rbind /media /media --options ro
‣ Building default image
‣ + sh -c 'trap '"'"'rm -rf /var/tmp/mkosi-var-tmp-3978212edd8b4fd4'"'"' EXIT && mkdir --mode 1777 /var/tmp/mkosi-var-tmp-3978212edd8b4fd4 && $0 "$@"' bwrap --ro-bind /home/talos/Desktop/debian-tools/usr /usr --unshare-net --die-with-parent --proc /proc --setenv SYSTEMD_OFFLINE 0 --tmpfs /tmp --unshare-ipc --dev /dev --bind /var/tmp/mkosi-var-tmp-3978212edd8b4fd4 /var/tmp --symlink usr/bin /bin --symlink usr/sbin /sbin --symlink usr/lib /lib --symlink usr/lib32 /lib32 --symlink usr/lib64 /lib64 --setenv PATH :/usr/bin:/usr/sbin --symlink ../proc/self/mounts /etc/mtab --ro-bind /home/talos/Desktop/debian-tools/etc/alternatives /etc/alternatives sh -c 'chmod 1777 /dev/shm && chmod 755 /etc && exec $0 "$@"' /usr/bin/ukify --version
‣ Could not find 'systemd-repart' which is required to build disk images.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mkosi/run.py", line 162, in uncaught_exception_handler
    yield
  File "/usr/lib/python3/dist-packages/mkosi/run.py", line 203, in fork_and_wait
    target()
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 3583, in <lambda>
    fork_and_wait(lambda: run_build(args, config))
                          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 3478, in run_build
    check_tools(config, Verb.build)
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 2190, in check_tools
    check_systemd_tool(config, "systemd-repart", version="254", reason="build disk images")
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 2169, in check_systemd_tool
    tool = check_tool(config, *tools, reason=reason, hint=hint)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 2157, in check_tool
    die(f"Could not find '{tools[0]}' which is required to {reason}.", hint=hint)
  File "/usr/lib/python3/dist-packages/mkosi/log.py", line 31, in die
    sys.exit(1)
SystemExit: 1
‣ + tput cnorm
‣ + tput smam
```

**Reply by daandemeyer** [reply]
> actually if I run `mkosi --tools-tree build` I get this:
> 
> ```
> usage: 
>   mkosi [options...] summary
>   mkosi [options...] build       [command line...]
>   mkosi [options...] shell       [command line...]
>   mkosi [options...] boot        [nspawn settings...]
>   mkosi [options...] qemu        [qemu parameters...]
>   mkosi [options...] ssh         [command line...]
>   mkosi [options...] journalctl  [command line...]
>   mkosi [options...] coredumpctl [command line...]
>   mkosi [options...] clean
>   mkosi [options...] serve
>   mkosi [options...] bump
>   mkosi [options...] genkey
>   mkosi [options...] documentation
>   mkosi [options...] help
>   mkosi -h | --help
>   mkosi --version
> mkosi: error: argument --tools-tree: expected one argument
> ```

You're on an older version of mkosi, the one packaged in Noble is already considered rather old, I would recommend running from git. You can try `mkosi --tools-tree=default build`. `mkosi dependencies` won't actually work on your version as it was added more recently.

**Reply by gritelli** [reply]
the output of `sudo mkosi --tools-tree=default build` is:
```
‣ Building default image
‣ Could not find 'systemd-repart' which is required to build disk images.
```
What version should I be running?
I am now on `mkosi 20.2` but I also tried 23.1-1 (I think) and got the same results.
Also tried the instructions to run from source, different errors, but still no build

**Reply by daandemeyer** [reply]
@gritelli Please try again to run from source, and if that doesn't work open a bug report and include the full output (add the `--debug` option to get debug output). I cannot help you with bugs on older versions except for telling you to run from source.

**Reply by gritelli** [reply]
bug opened. Hoepfully there are enough info for you to figure out what the issue is.

**Reply by unkownorca** [reply]
Hello im having the same issue rn with the systemd-repart I have literally the same version of the host and I still can't find a solution did you manage to get it to work thanks!

**Comment by ferrar12121987** [comment]
Hey I am experiencing the same issue, I would appreciate the support as I am currently doing a build tool down-select and MKOSI looks promising. Please see bellow:

```
ubuntu@ip-172-31-5-27:~/test_mkosi/mkosi-for-ubuntu$ ls
README.md  debian-tools  mkosi.default  mkosi.extra  mkosi.postinst
ubuntu@ip-172-31-5-27:~/test_mkosi/mkosi-for-ubuntu$ sudo mkosi --tools-tree=default --debug build
‣ Including configuration file /tmp/tmplepaj5yn/mkosi-tools/mkosi.conf
‣ Including configuration file /tmp/tmplepaj5yn/mkosi-tools/mkosi.conf.d/10-debian-ubuntu.conf
‣ Including configuration file /tmp/tmph008dgb0/mkosi-tools/mkosi.conf
‣ Including configuration file /tmp/tmph008dgb0/mkosi-tools/mkosi.conf.d/10-debian-ubuntu.conf
‣ + mount --make-rslave /
‣ + mount --rbind /usr /usr --options ro
‣ + mount --rbind /etc /etc --options ro
‣ + mount --rbind /opt /opt --options ro
‣ + mount --rbind /boot /boot --options ro
‣ + mount --rbind /media /media --options ro
‣ Building default image
‣ + sh -c 'trap '"'"'rm -rf /var/tmp/mkosi-var-tmp-052da6d039d94485'"'"' EXIT && mkdir --mode 1777 /var/tmp/mkosi-var-tmp-052da6d039d94485 && $0 "$@"' bwrap --ro-bind /home/ubuntu/test_mkosi/mkosi-for-ubuntu/debian-tools/usr /usr --unshare-net --die-with-parent --proc /proc --setenv SYSTEMD_OFFLINE 0 --tmpfs /tmp --unshare-ipc --dev /dev --bind /var/tmp/mkosi-var-tmp-052da6d039d94485 /var/tmp --symlink usr/bin /bin --symlink usr/sbin /sbin --symlink usr/lib /lib --symlink usr/lib32 /lib32 --symlink usr/lib64 /lib64 --setenv PATH :/usr/bin:/usr/sbin --symlink ../proc/self/mounts /etc/mtab --ro-bind /home/ubuntu/test_mkosi/mkosi-for-ubuntu/debian-tools/etc/alternatives /etc/alternatives sh -c 'chmod 1777 /dev/shm && chmod 755 /etc && exec $0 "$@"' /usr/bin/ukify --version
‣ Could not find 'systemd-repart' which is required to build disk images.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/mkosi/run.py", line 162, in uncaught_exception_handler
    yield
  File "/usr/lib/python3/dist-packages/mkosi/run.py", line 203, in fork_and_wait
    target()
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 3583, in <lambda>
    fork_and_wait(lambda: run_build(args, config))
                          ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 3478, in run_build
    check_tools(config, Verb.build)
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 2190, in check_tools
    check_systemd_tool(config, "systemd-repart", version="254", reason="build disk images")
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 2169, in check_systemd_tool
    tool = check_tool(config, *tools, reason=reason, hint=hint)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/mkosi/__init__.py", line 2157, in check_tool
    die(f"Could not find '{tools[0]}' which is required to {reason}.", hint=hint)
  File "/usr/lib/python3/dist-packages/mkosi/log.py", line 31, in die
    sys.exit(1)
SystemExit: 1
‣ + tput cnorm
‣ + tput smam
ubuntu@ip-172-31-5-27:~/test_mkosi/mkosi-for-ubuntu$ 
```

**Reply by daandemeyer** [reply]
@ferrar12121987 As explained above, `mkosi dependencies` will output a list of packages that are dependencies. You seem to be running an older version of mkosi that doesn't have this verb. Given the fast rate of development of mkosi, I recommend you pin the latest commit of mkosi and run from that to make sure you get the best experience. We'll do another release soon but that won't end up in the Ubuntu LTS of course.
